runtime/lua/vim/lsp/util.lua

```diff
@@ -474,7 +474,20 @@ function M.apply_text_edits(text_edits, bufnr, offset_encoding)
       end
       has_eol_text_edit = true
     end
-    api.nvim_buf_set_text(bufnr, e.start_row, e.start_col, e.end_row, e.end_col, e.text)
+
+    local last_line = get_line(bufnr, e.end_row)
+
+    -- If the replacement is over full lines it is safer to use api.nvim_buf_set_lines
+    if e.end_col >= #last_line and e.start_col == 0 then
+      -- When replacing whole lines there seems to always be an extra newline at the end of the new
+      -- text. So we remove it.
+      if #e.text > 1 and e.text[#e.text] == '' then
+        table.remove(e.text, #e.text)
```

Comment:
Isn't this kinda similar to what is happening above in line 464 and 471?

Reply:
Similar yes but not the same. Those cases looks at the line ranges in the buffer that are to be replaced. I'm looking at the column ranges in the buffer to know if they represent a full set on lines (i.e. if the first line start at the beginning and the last lines ends at the end). The inner if looks at the new text that is to be inserted and ensures that there isn't an extra empty line in there. That happens because the replacement text in these cases include the newline at the end of the block but neovim does replacement in a way that it wants to handle the linebreaks by itself. Another solution would be if we could do the replacement in a way that included the new line in the buffer and replace with the exact content of `text_edit.newText` without splitting it to several lines manually. I tried to find alternatives to `nvim_buf_set_text` that can do that but I was unsuccessful.